openedx_learning/apps/authoring/publishing/api.py

```diff
@@ -644,87 +782,402 @@ def _add_to_existing_draft_change_log(
+            untrusted_record_id_set,
+        )
+
+    _bulk_update_hashes(record_cls, records_that_need_hashes)
+
+
+def _bulk_update_hashes(model_cls, records):
+    """
+    bulk_update using the model class (PublishLogRecord or DraftChangeLogRecord)
+
+    The only reason this function exists is because mypy 1.18.2 throws an
+    exception in validate_bulk_update() during "make quality" checks otherwise
+    (though curiously enough, not when that same version of mypy is called
+    directly). Given that I'm writing this on the night before the Ulmo release
+    cut, I'm not really interested in tracking down the underlying issue.
+
+    The lack of type annotations on this function is very intentional.
+    """
+    model_cls.objects.bulk_update(records, ['dependencies_hash_digest'])
+
+
+def hash_for_log_record(
+    record: DraftChangeLogRecord | PublishLogRecord,
+    record_ids_to_hash_digests: dict,
+    record_ids_to_live_deps: dict,
+    untrusted_record_id_set: set | None,
+) -> str:
+    """
+    The hash digest for a given change log record.
+
+    Note that this code is a little convoluted because we're working hard to
+    minimize the number of database requests. All the data we really need could
+    be derived from just the record itself, but at a far higher cost.
+
+    When we say "dependency", we mean when one or more PublishableEntities are
+    dependencies of a PublishableEntityVersion. A change in the Draft or
+    Published state of the PublishableEntity causes an implicit change to the
+    PublishableEntityVersion. The common case we have at the moment is when a
+    container type like a Unit has Components as dependencies, i.e. the
+    UnitVersion specifies Component Entities as dependencies.
+
+    This means that the dependencies_hash_digest that summarizes the total state
+    of a PublishableEntity depends on:
+
+    1. The definition of the current draft/published version of that entity.
+    2. The current draft/published versions of all dependencies.
+
+    Which is why it makes sense to capture in a log record, since
+    PublishLogRecords or DraftChagneLogRecords are created whenever one of the
+    above two things changes.
+
+    Here are the possible scenarios, including edge cases:
+
+    Soft-deletions
+      If the record.new_version is None, that means we've just soft-deleted
+      something (or published the soft-delete of something). We adopt the
+      convention that if something is soft-deleted, its dependencies_hash_digest
+      is reset to the default value of ''.
+
+    EntityVersions with no dependencies
+      If record.new_version has no dependencies, dependencies_hash_digest is
+      likewise set to the default value of ''. This will be the most common
+      case.
+
+    EntityVersions with dependencies
+      If an EntityVersion has dependencies, we calculate the dependency hash
+      digest by taking a hash of all the UUIDs of the active version. [ fill in
+      more later]
```

Comment:
here, let me try to fill it in:
```suggestion
    EntityVersions with dependencies
      If an EntityVersion has dependencies, then its draft/published state
      hash is based on the concatenation of, for each non-deleted dependency:
        (i)  the dependency's draft/published EntityVersion primary key, and
        (ii) the dependency's own draft/published state hash, recursively re-
             calculated if necessary.
```